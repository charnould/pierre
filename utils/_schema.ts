import { z } from 'zod'

/**
 * Represents a User schema definition using Zod.
 *
 * This schema validates the structure of a user object with the following properties:
 *
 * - `email`: A string that is trimmed and converted to lowercase.
 * - `role`: An enumerated string that can be one of 'administrator', 'contributor', or 'collaborator'.
 * - `config`: A string that is trimmed.
 * - `password_hash`: A string representing the hashed password of the user.
 */
export const User = z.object({
  email: z.string().trim().toLowerCase(),
  role: z.enum(['administrator', 'contributor', 'collaborator']).catch('collaborator'),
  config: z.string().trim(),
  password_hash: z.string()
})

/**
 * Represents a parsed user schema that extends the base `User` schema
 * by adding a `config` property. The `config` property is an array
 * of strings, allowing for additional configuration options.
 *
 * This schema can be used to validate and type-check user data
 * with additional configuration details.
 */
export const Parsed_User = User.extend({ config: z.array(z.string()) })

//
// Incoming SMS parsing schema
export const SMS = z
  .object({
    role: z.string(),
    config: z.string(),
    conv_id: z.string(),
    phone: z.string().nullable(),
    to: z.string(),
    content: z.string().trim()
  })
  .or(z.null())

//
// `./assets/Config` schema
export const Config = z
  .object({
    id: z.string(),
    display: z.string(),
    show: z.array(z.string()),

    custom_data: z.object({ format: z.function() }).or(z.object({})),

    api: z
      .array(
        z.object({
          key: z.enum(['WEBHOOK_KEY_1', 'WEBHOOK_KEY_2', 'WEBHOOK_KEY_3']),
          url: z.string(), // URL check is made elsewhere
          format: z.function().args(
            z.object({
              custom_data: z.array(z.string()),
              content: z.string(),
              role: z.string()
            })
          )
        })
      )
      .default([]),

    models: z
      .object({
        augment_with: z.string(),
        rerank_with: z.string(),
        answer_with: z.string()
      })
      .default({
        augment_with: "openai('gpt-4o-mini-2024-07-18')",
        rerank_with: "openai('gpt-4o-mini-2024-07-18')",
        answer_with: "openai('gpt-4o-mini-2024-07-18')"
      }),
    phone: z.string().nullable(),
    protected: z.boolean(),
    knowledge: z.object({
      community: z.boolean(),
      proprietary: z.object({
        public: z.boolean(),
        private: z.boolean()
      })
    }),
    audience: z.string(),
    persona: z.string(),
    greeting: z.array(z.string()),
    examples: z.array(z.string()),
    disclaimer: z.string().nullable()
  })
  .strict()

//
// Reflects datastore database schema
export const Reply = z.object({
  // Globals
  conv_id: z.string(),
  config: z.string().or(Config),
  role: z.enum(['assistant', 'user', 'system']).default('user'),
  timestamp: z.string().datetime({ offset: true }).nullish().default(null),
  content: z.string(),
  // Metadata
  metadata: z
    .object({
      // User
      user: z.string().trim().toLowerCase().nullish().default(null),
      // Topics
      topics: z.string().trim().toLowerCase().nullish().default(null),
      // Evaluation + satisfaction
      evaluation: z
        .object({
          // CSAT
          customer: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // Social housing organization satisfaction
          organization: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // AI generated customer satisfaction (CSAT)
          ai: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({})
        })
        .default({}),
      // LLM usage
      tokens: z
        .object({
          // cache : z.number().nullish().default(null),
          prompt: z.number().nullish().default(null),
          completion: z.number().nullish().default(null),
          total: z.number().nullish().default(null)
        })
        .default({})
    })
    .default({})
})

//
// Structured JSON LLM must output for each request
export const Augmented_Query = z.object({
  lang: z.string(),
  contains_profanity: z.boolean(),
  standalone_questions: z.array(z.string()),
  bm25_keywords: z.array(z.string()),
  search_queries: z.array(z.string())
})

//
// AI Context
export const AIContext = Reply.extend({ query: Augmented_Query.nullable().default(null) })
  .merge(
    z.object({
      chunks: z
        .object({
          community: z.array(z.string()).nullish().default([]),
          private: z.array(z.string()).nullish().default([]),
          public: z.array(z.string()).nullish().default([])
        })
        .default({
          community: [],
          private: [],
          public: []
        }),
      custom_data: z.object({ raw: z.array(z.string()), transformed: z.string().default('') }),
      conversation: z
        .array(
          z.object({
            role: z.enum(['assistant', 'user', 'system']),
            content: z.string()
          })
        )
        .default([])
    })
  )
  .refine(async (c) => {
    // Change config name for config content
    if (typeof c.config === 'string') {
      c.config = (await import(`../assets/${c.config}/config`)).default
    }

    // Format data to be the one wanted by API
    if (typeof c.config !== 'string' && 'format' in c.config.custom_data) {
      if (
        Array.isArray(c.custom_data.raw) &&
        c.custom_data.raw.length === 1 &&
        c.custom_data.raw[0] === ''
      ) {
        c.custom_data.transformed = ''
      } else {
        c.custom_data.transformed = c.config.custom_data.format(c.custom_data.raw) as string
      }
    }

    return true
  })

//
//
export type SMS = z.infer<typeof SMS>
export type User = z.infer<typeof User>
export type Parsed_User = z.infer<typeof Parsed_User>
export type Reply = z.infer<typeof Reply>
export type Config = z.infer<typeof Config>
export type AIContext = z.infer<typeof AIContext>
export type Augmented_Query = z.infer<typeof Augmented_Query>
