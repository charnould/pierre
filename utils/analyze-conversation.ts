import type { ModelMessage } from 'ai'
import { SQL } from 'bun'
import dedent from 'dedent'
import type { Reply } from './_schema'
import { extract_tag_value } from './augment-query'
import { generate_text } from './generate-output'
import { get_conversation, save_topic, score } from './handle-conversation'

const sql = new SQL(`sqlite:datastores/${Bun.env.SERVICE}/datastore.sqlite`)

/**
 * Scores conversations using an AI model.
 *
 * This function retrieves conversation IDs that have no score from the
 * database, then for each conversation, it retrieves the conversation content,
 * uses a language model to determine the score and reasoning, and saves the
 * score back into the database.
 *
 * The scoring process involves:
 * 1. Retrieving all conversation content.
 * 2. Using a language model to evaluate the conversation and determine a score.
 * 3. Saving the score and reasoning back into the database.
 *
 * The score is determined based on how well the AI understood and addressed the
 * user's intent, with the following scale:
 * - 0: Completely failed
 * - 1: Somewhat relevant but major issues
 * - 2: Mostly good but minor flaws
 * - 3: Perfectly understood and responded
 *
 * If the conversation contains only one message, a specific score and comment
 * are assigned.
 *
 * @returns {Promise<void>} A promise that resolves when the scoring process is complete.
 */
export const score = async (): Promise<void> => {
  // Get the `conv_id` of conversations that have no score
  const sql = new SQL(`sqlite:datastores/${Bun.env.SERVICE}/datastore.sqlite`)
  let conv_ids_missing_score = await sql`
    SELECT DISTINCT
      conv_id
    FROM
      telemetry
    WHERE
      json_extract (metadata, '$.evaluation.ai.score') IS NULL;
  `
  conv_ids_missing_score = conv_ids_missing_score.map((row) => row.conv_id)

  // For each conversation:
  // 1. Retrieve all conversation content.
  // 2. Use an LLM to determine the topic.
  // 3. Save the topic back into the database.
  for await (const conv_id of conv_ids_missing_score) {
    const conversation = (await get_conversation(conv_id)) as Reply[]
    const core_messages = conversation.map(({ role, content }) => ({
      role,
      content
    })) as ModelMessage[]

    let score: number | string | null | boolean
    let comment: number | string | null | boolean

    if (core_messages.length === 1) {
      score = -1
      comment = "PIERRE n'a pas répondu à l'utilisateur."
    } else {
      const messages: ModelMessage[] = [
        ...core_messages,
        {
          role: 'assistant',
          content: dedent`
              Evaluate the preceding AI-user conversation and assign a quality score based on how well the AI understood and addressed the user's intent. Follow these precise steps:
              
              1. Carefully read the entire conversation, noting key requests, questions, and AI responses.
              
              2. Analyze whether the AI:
              - Correctly interpreted the user's primary intent and any implicit needs
              - Provided accurate, relevant, and complete information
              - Structured the response appropriately for the context
              - Avoided unnecessary tangents or omissions
              
              3. Assign ONE score from this scale:
              0: Failed completely (misunderstood the request or provided incorrect/harmful information)
              1: Partially addressed the request (major gaps, inaccuracies, or misalignments with intent)
              2: Generally good response (minor improvements possible in accuracy, completeness, or relevance)
              3: Excellent response (perfectly understood intent and provided optimal assistance)
              
              4. Format your response EXACTLY as follows:
              <score>X</score>
              <reasoning>Your 20-75 word justification in French, explaining specifically why you assigned this score, with concrete examples from the conversation.</reasoning>
              
              Ensure proper XML syntax with matching opening and closing tags. The score must be a single digit (0, 1, 2, or 3) between the score tags.`
        }
      ]

      const config = (await import('../assets/default/config')).default
      const model = config.models
      const answer = await generate_text({
        model: model,
        messages: messages,
        max_tokens: 200
      })

      score = extract_tag_value(answer, 'score', null)
      comment = extract_tag_value(answer, 'reasoning', null)
    }

    await score({
      conv_id: conv_id,
      scorer: 'ai',
      score: score,
      comment: comment
    })
  }

  console.info('✅ AI score assignment completed')
  return
}

/**
 * Assigns a topic to conversations that have no assigned topic using AI.
 *
 * This function performs the following steps:
 * 1. Retrieves the `conv_id` of conversations that have no assigned topic from the database.
 * 2. For each conversation, retrieves all conversation content.
 * 3. Uses a language model to determine the topic of the conversation.
 * 4. Saves the determined topic back into the database.
 *
 * The predefined categories for topics are:
 * - 'rent': The discussion pertains to rent payments.
 * - 'life': The conversation is about daily living in the residence.
 * - 'insurance': The topic concerns a home insurance.
 * - 'moving': The discussion revolves around relocation.
 * - 'technical': The conversation involves maintenance requests.
 * - 'contact': The user is inquiring about ways to contact the lessor.
 * - 'chatbot': The user is asking about the chatbot.
 * - 'multiple': More than one of the above categories is equally relevant.
 * - 'unknown': None of the above categories accurately describe the conversation.
 *
 * @returns {Promise<void>} A promise that resolves when the operation is complete.
 */
export const topicize = async (): Promise<void> => {
  // Get the `conv_id` of conversations that have no assigned topic
  let conv_ids_missing_topic = await sql`
    SELECT DISTINCT
      conv_id
    FROM
      telemetry
    WHERE
      json_extract (metadata, '$.topics') IS NULL;
  `
  conv_ids_missing_topic = conv_ids_missing_topic.map((row) => row.conv_id)

  // For each conversation:
  // 1. Retrieve all conversation content.
  // 2. Use an LLM to determine the topic.
  // 3. Save the topic back into the database.
  for await (const conv_id of conv_ids_missing_topic) {
    const conversation = (await get_conversation(conv_id)) as Reply[]
    const core_messages = conversation.map(({ role, content }) => ({
      role,
      content
    })) as ModelMessage[]

    const messages: ModelMessage[] = [
      ...core_messages,
      {
        role: 'assistant',
        content: dedent`
        
        Analyze the full preceding conversation and classify its primary topic into exactly one of the following predefined categories:
        
        - 'rent': The discussion pertains to rent payments, due dates, amounts, late fees, or related financial matters.
        - 'life': The conversation is about daily living in the residence, neighbors, amenities, social aspects, or general experiences.
        - 'insurance': The topic concerns a home insurance certificate, its submission, validity, or related requirements.
        - 'moving': The discussion revolves around relocation, move-in/move-out dates, lease transfers, or associated logistics.
        - 'technical': The conversation involves maintenance requests, repairs, internet issues, heating problems, or any technical difficulties in the residence.
        - 'contact': The user is inquiring about ways to contact the lessor, landlord, or property management.
        - 'social_housing': The user is asking about obtaining social housing, eligibility, application procedures, waiting lists, or related topics.
        - 'chatbot': The user is asking about the chatbot, how responses are generated, or related AI-specific questions.
        - 'multiple': More than one of the above categories is equally relevant.
        - 'unknown': None of the above categories accurately describe the conversation.
        
        Response Instructions:
        - Return only the exact category name (e.g., rent).
        - Do not include explanations, extra words, formatting, or any additional output.
        - Ensure the classification is 100% consistent and deterministic based on the given definitions.
        
        Your classification:`
      }
    ]

    const topic = (await generate_text({ messages: messages, max_tokens: 50 })).toLowerCase().trim()

    const topics = [
      'rent',
      'life',
      'insurance',
      'moving',
      'technical',
      'contact',
      'social_housing',
      'chatbot',
      'multiple',
      'unknown'
    ]

    if (topics.includes(topic)) await save_topic({ conv_id: conv_id, topic: topic })
  }

  console.info('✅ AI topic assignment completed')
  return
}
