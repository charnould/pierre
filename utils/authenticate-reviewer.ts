import type { Context, Next } from 'hono'
import { getSignedCookie } from 'hono/cookie'

export const authenticate = async (c: Context, next: Next) => {
  const cookie = await getSignedCookie(c, Bun.env.AUTH_SECRET as string, 'pierre')
  if (cookie === undefined) c.set('is_auth', false)
  else c.set('is_auth', true)
  await next()
}
