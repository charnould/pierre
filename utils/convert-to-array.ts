/**
 * Converts a given text string into an array of strings by performing several transformations:
 * - Removes any content wrapped in `<think>` tags.
 * - Removes escaped double quotes (`\"`) and regular double quotes (`"`).
 * - Splits the string by the pipe character (`|`).
 * - Trims whitespace and removes newline characters from each resulting line.
 * - Filters out any empty strings from the result.
 *
 * @note This function is tested in tests/utils/convert-to-array.test.ts
 * @param text - The input string to be converted.
 * @returns An array of non-empty, cleaned strings.
 */
export const convert_to_array = (text: string): string[] =>
  text
    .replace(/<think>\s*(.*?)\s*<\/think>/, ' ')
    .replace(/\\"/g, '')
    .replace(/"/g, '')
    .split('|')
    .map((line) => line.replace(/\n/g, '').trim())
    .filter((line) => line.length > 0)
