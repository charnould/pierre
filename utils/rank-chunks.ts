import { createAnthropic } from '@ai-sdk/anthropic'
import { createCerebras } from '@ai-sdk/cerebras'
import { createCohere } from '@ai-sdk/cohere'
import { createGoogleGenerativeAI } from '@ai-sdk/google'
import { createGroq } from '@ai-sdk/groq'
import { createMistral } from '@ai-sdk/mistral'
import { createOpenAI } from '@ai-sdk/openai'
import { createTogetherAI } from '@ai-sdk/togetherai'
import dedent from 'dedent'
import _ from 'lodash'
import * as prettier from 'prettier'
import { z } from 'zod'
import type { AIContext } from './_schema'
import { today_is } from './generate-answer'
import { generate_text } from './generate-output'
import type { Vector_Search_Result } from './search-by-vectors'

/**
 * Ranks and filters chunks based on their relevance scores.
 *
 * @param vect_data - Array of vector search results.
 * @param bm25_data - Array of BM25 search results.
 * @param context - AI context used for scoring chunks.
 * @returns A promise that resolves to the most relevant chunks.
 *
 * The function performs the following steps:
 * 1. Flattens and combines vector and BM25 search results.
 * 2. Scores each chunk using the provided context.
 * 3. Orders the chunks by their scores in descending order.
 * 4. Filters and returns the most relevant chunks.
 *
 * If the `DEBUG` environment variable is set, the function will also write
 * intermediate data to JSON files for debugging purposes.
 */
export const rank_chunks = async (
  vect_data: Vector_Search_Result[],
  bm25_data: Vector_Search_Result[],
  context: AIContext
) => {
  // Flatten chunks and combine them
  const vect_chunks = flatten_vector_searches(vect_data)
  const bm25_chunks = _.differenceBy(flatten_vector_searches(bm25_data), vect_chunks, 'chunk_hash')
  const chunks = [...vect_chunks, ...bm25_chunks]

  // Score chunks...
  const chunk_scores = await Promise.all(chunks.map((chunk) => score_chunk(context, chunk)))

  // Retrive chunk content and its source (community, private, public)...
  const scored_chunks = _.orderBy(
    chunks.map((chunk, index) => ({
      score: chunk_scores[index].score,
      reasoning: chunk_scores[index].reasoning,
      chunk_text: chunk.chunk_text,
      source: chunk.source
    })),
    ['score'],
    ['desc']
  )

  // Finally keep only the most relevant ones...
  const relevant_chunks = pick_relevant_chunks(scored_chunks)

  // For debugging purpose only
  if (Bun.env.DEBUG) {
    const print = async (
      data:
        | {
            chunk_hash: string
            chunk_text: string
            distance: number
            source: string
          }[]
        | {
            score: number
            reasoning: string | null
            chunk_text: string
            source: string
          }[]
        | { community: string[]; private: string[]; public: string[] }
    ) => await prettier.format(JSON.stringify(data), { parser: 'json' })
    Bun.write('__temp__/1. search-vector.json', await print(vect_chunks))
    Bun.write('__temp__/2. search-bm25.json', await print(bm25_chunks))
    Bun.write('__temp__/3. chunks-combined.json', await print(chunks))
    Bun.write('__temp__/4. chunks-scored.json', await print(scored_chunks))
    Bun.write('__temp__/5. chunks-relevant.json', await print(relevant_chunks))
  }

  return Relevant_Chunks.parse(relevant_chunks)
}

//
//
//
//
// Typescrit Types
//
//
//
//

// prettier-ignore
// biome-ignore format: readability
const Flatten_Chunk = z.object({
  chunk_hash  : z.string(),
  distance    : z.number(),
  chunk_text  : z.string(),
  source      : z.string()
})

// prettier-ignore
// biome-ignore format: readability
const Relevant_Chunks = z.object({
  community     : z.array(z.string()).default([]),
  private       : z.array(z.string()).default([]),
  public        : z.array(z.string()).default([])
})

export type Flatten_Chunk = z.infer<typeof Flatten_Chunk>
export type Relevant_Chunks = z.infer<typeof Relevant_Chunks>

//
//
//
//
//
// Helpers
//
//
//
//
//

/**
 * Flattens and ranks vector search results.
 *
 * This function takes an array of search results, flattens the nested
 * structure, groups the results by their source, and then ranks the chunks
 * within each group by their distance. It returns the top 50 closest chunks for
 * each source.
 *
 * @param {Vector_Search_Result[]} data - The array of search results to be processed.
 * @returns {Flatten_Chunk[]} - The flattened and ranked array of chunks.
 */
export const flatten_vector_searches = (data: Vector_Search_Result[]): Flatten_Chunk[] =>
  _.chain(data)
    .flatMap((o) =>
      Object.entries(o).flatMap(([source, items]) => {
        if (Array.isArray(items)) {
          return items.map((item) => ({
            ...item,
            source
          }))
        }
        return []
      })
    )
    .groupBy('source')
    .mapValues((entries) =>
      _(entries)
        .groupBy('chunk_hash')
        .map((group) => _.minBy(group, 'distance'))
        .compact()
        .orderBy('distance')
        .take(25)
        .value()
    )
    .values()
    .flatMap()
    .value()

/**
 * Picks the most relevant chunks from the provided scores.
 *
 * @param scores - An array of objects containing score, chunk_text, and source.
 * @returns An object with keys as sources and values as arrays of the top 5 chunk_texts
 *          with scores greater than or equal to 500, ordered by score in descending order.
 *          Defaults to empty arrays for sources 'public', 'private', and 'community'.
 */
export const pick_relevant_chunks = (
  scores: {
    score: number
    chunk_text: string
    source: string
  }[]
): Relevant_Chunks =>
  _.chain(scores)
    .groupBy('source')
    .mapValues((array) =>
      _.chain(array)
        .orderBy('score', 'desc')
        .filter((o) => o.score >= 500)
        .take(5)
        .map('chunk_text')
        .value()
    )
    .defaults({ public: [], private: [], community: [] })
    .value()

/**
 * Scores a given chunk of text based on its relevance and clarity in answering the user's final intent.
 *
 * @note This function is manually tested using `bun eva:reranker` command.
 * @param context - The AI context containing the conversation history and configuration.
 * @param chunk - The chunk of text to be evaluated.
 * @returns A promise that resolves to the score of the chunk.
 *
 * The function uses an LLM to generate a score based on multiple following criteria.
 * The score is enclosed within <score> tags in the response for an easy retrieval.
 * CAUTION: if prompt ask only for a score, reasoning quality will be low!
 */
export const score_chunk = async (context: AIContext, chunk: Flatten_Chunk) => {
  const openai = createOpenAI({ compatibility: 'strict' })
  const google = createGoogleGenerativeAI()
  const anthropic = createAnthropic()
  const mistral = createMistral()
  const cohere = createCohere()
  const togetherai = createTogetherAI()
  const groq = createGroq()
  const cerebras = createCerebras()

  const score = await generate_text({
    messages: [
      ...context.conversation,
      {
        role: 'system',
        content: dedent`
        
        You are an expert in semantic relevance evaluation and contextual understanding. Your task is to analyze the entire conversation between the user and the agent to determine whether the provided chunk effectively answers the user’s final intent or standalone question.

        Today is ${today_is()}.
        
        You are evaluating the following chunk:

        <chunk>
        ${chunk.chunk_text}
        </chunk>
        
        **Evaluation Criteria**
        
        1. Answer Presence – Does the chunk contain the answer to the user’s final intent?
        2. Relevance – Is the answer directly related to the user’s question, including implicit meaning?
        
        **Validation Check**
        
        - Correct Reference – If the user refers to a specific building but the chunk discusses a different one, score 0, unless the response is still broadly relevant.
        - Hidden Answer – If only a small part of the chunk answers the question, score based on that portion.
        
        **Scoring Scale**

        - 1000 (Perfect Match) – Accurately answers the final user intent.
        - 1-999 (Partial Match) – The response is relevant but lacks precision, completeness, or clarity.
        - 0 (No Match) – No meaningful connection to the user’s intent.
        
        **Think carefully and output the score first, within <score> tags, followed by your reasoning.**`
      }
    ],
    model: context.config.models.rerank_with,
    max_tokens: 300
  })

  return extract_score_and_reasoning(score)
}

/**
 * Extracts a numerical score from a given text string.
 *
 * @note This function is tested in tests/unit/utils/extract-score.test.ts
 * @param text - The text string from which to extract the score. It can be a string, null, or undefined.
 * @returns The extracted score as a number. If the text is not a string or if no score is found, returns 0.
 */
export const extract_score_and_reasoning = (text: string | null | undefined) => {
  if (typeof text !== 'string' || text === '') {
    return { score: 0, reasoning: null }
  }

  const match = text.match(/<score>\s*(\d+)\s*<\/score>/)
  const score = match ? Number.parseFloat(match[1]) : 0
  return { score: score, reasoning: text }
}
