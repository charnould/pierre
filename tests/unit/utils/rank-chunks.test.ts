import { expect, it } from 'bun:test'
import { flatten_vector_searches, pick_relevant_chunks } from '../../../utils/rank-chunks'

const input = [
  {
    community: [
      {
        chunk_hash: 'hash_a',
        entity_hash: 'none',
        distance: 0.01,
        chunk_text: 'text_a'
      },
      {
        chunk_hash: 'hash_b',
        entity_hash: 'none',
        distance: 0.02,
        chunk_text: 'text_b'
      },
      {
        chunk_hash: 'hash_c',
        entity_hash: 'none',
        distance: 0.03,
        chunk_text: 'text_c'
      }
    ],
    private: [
      {
        chunk_hash: 'hash_1',
        entity_hash: 'entity_1',
        distance: 0.01,
        chunk_text: 'text_1'
      },
      {
        chunk_hash: 'hash_2',
        entity_hash: 'entity_1',
        distance: 0.02,
        chunk_text: 'text_2'
      },
      {
        chunk_hash: 'hash_3',
        entity_hash: 'entity_2',
        distance: 0.03,
        chunk_text: 'text_3'
      }
    ],
    public: [
      {
        chunk_hash: 'hash_4',
        entity_hash: 'entity_3',
        distance: 0.02,
        chunk_text: 'text_4'
      },
      {
        chunk_hash: 'hash_5',
        entity_hash: 'entity_4',
        distance: 0.03,
        chunk_text: 'text_5'
      },
      {
        chunk_hash: 'hash_6',
        entity_hash: 'entity_5',
        distance: 0.04,
        chunk_text: 'text_6'
      }
    ]
  },
  {
    community: [
      {
        chunk_hash: 'hash_d',
        entity_hash: 'none',
        distance: 0.03,
        chunk_text: 'text_d'
      },
      {
        chunk_hash: 'hash_e',
        entity_hash: 'none',
        distance: 0.02,
        chunk_text: 'text_e'
      },
      {
        chunk_hash: 'hash_f',
        entity_hash: 'none',
        distance: 0.01,
        chunk_text: 'text_f'
      }
    ],
    private: [
      {
        chunk_hash: 'hash_7',
        entity_hash: 'entity_6',
        distance: 0.01,
        chunk_text: 'text_7'
      },
      {
        chunk_hash: 'hash_8',
        entity_hash: 'entity_6',
        distance: 0.02,
        chunk_text: 'text_8'
      },
      {
        chunk_hash: 'hash_9',
        entity_hash: 'entity_6',
        distance: 0.03,
        chunk_text: 'text_9'
      }
    ],
    public: [
      {
        chunk_hash: 'hash_4',
        entity_hash: 'entity_1',
        distance: 0.001,
        chunk_text: 'text_1'
      },
      {
        chunk_hash: 'hash_5',
        entity_hash: 'entity_1',
        distance: 0.2,
        chunk_text: 'text_2'
      },
      {
        chunk_hash: 'hash_6',
        entity_hash: 'entity_2',
        distance: 0.003,
        chunk_text: 'text_3'
      }
    ]
  }
]

//
//
//
//
//
//
//
const chunks = [
  {
    score: 100,
    chunk_text: 'chunk_75',
    source: 'public'
  },
  {
    score: 1000,
    chunk_text: 'chunk_76',
    source: 'public'
  },
  {
    score: 0,
    chunk_text: 'chunk_78',
    source: 'public'
  },
  {
    score: 630,
    chunk_text: 'chunk_77',
    source: 'private'
  },
  {
    score: 0,
    chunk_text: 'chunk_1100',
    source: 'private'
  },
  {
    score: 600,
    chunk_text: 'chunk_1200',
    source: 'private'
  },
  {
    score: 0,
    chunk_text: 'chunk_1200',
    source: 'public'
  },
  {
    score: 800,
    chunk_text: 'chunk_1300',
    source: 'community'
  }
]

//
//
//
//
//
//
//

it('should flatten chunks 1', () => expect(flatten_vector_searches(input)).toMatchSnapshot())

it('should keep only relevant', () =>
  expect(pick_relevant_chunks(chunks)).toStrictEqual({
    private: ['chunk_77', 'chunk_1200'],
    public: ['chunk_76'],
    community: ['chunk_1300']
  }))
