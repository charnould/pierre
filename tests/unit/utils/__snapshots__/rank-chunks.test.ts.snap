// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`should flatten chunks 1 1`] = `
[
  {
    "chunk_hash": "hash_a",
    "chunk_text": "text_a",
    "distance": 0.01,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_f",
    "chunk_text": "text_f",
    "distance": 0.01,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_b",
    "chunk_text": "text_b",
    "distance": 0.02,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_e",
    "chunk_text": "text_e",
    "distance": 0.02,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_c",
    "chunk_text": "text_c",
    "distance": 0.03,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_d",
    "chunk_text": "text_d",
    "distance": 0.03,
    "entity_hash": "none",
    "source": "community",
  },
  {
    "chunk_hash": "hash_1",
    "chunk_text": "text_1",
    "distance": 0.01,
    "entity_hash": "entity_1",
    "source": "proprietary",
  },
  {
    "chunk_hash": "hash_7",
    "chunk_text": "text_7",
    "distance": 0.01,
    "entity_hash": "entity_6",
    "source": "proprietary",
  },
  {
    "chunk_hash": "hash_2",
    "chunk_text": "text_2",
    "distance": 0.02,
    "entity_hash": "entity_1",
    "source": "proprietary",
  },
  {
    "chunk_hash": "hash_8",
    "chunk_text": "text_8",
    "distance": 0.02,
    "entity_hash": "entity_6",
    "source": "proprietary",
  },
  {
    "chunk_hash": "hash_3",
    "chunk_text": "text_3",
    "distance": 0.03,
    "entity_hash": "entity_2",
    "source": "proprietary",
  },
  {
    "chunk_hash": "hash_9",
    "chunk_text": "text_9",
    "distance": 0.03,
    "entity_hash": "entity_6",
    "source": "proprietary",
  },
]
`;
